src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp

```diff
@@ -1041,25 +1043,78 @@ void SinglePhaseBase::applySourceFluxBC( real64 const time_n,
+        
+        arrayView2d< real64 const > const enthalpy = fluid.enthalpy();
+        arrayView2d< real64 const > const dEnthalpy_dTemperature = fluid.dEnthalpy_dTemperature();
+        arrayView2d< real64 const > const dEnthalpy_dPressure    = fluid.dEnthalpy_dPressure();
+        forAll< parallelDevicePolicy<> >( targetSet.size(), [sizeScalingFactor,
+                                                             targetSet,
+                                                             rankOffset,
+                                                             ghostRank,
+                                                             dofNumber,
+                                                             enthalpy,
+                                                             dEnthalpy_dTemperature,
+                                                             dEnthalpy_dPressure,
+                                                             rhsContributionArrayView,
+                                                             localRhs,
+                                                             localMatrix] GEOS_HOST_DEVICE ( localIndex const a )
         {
-          return;
-        }
+          // we need to filter out ghosts here, because targetSet may contain them
+          localIndex const ei = targetSet[a];
+          if( ghostRank[ei] >= 0 )
+          {
+            return;
+          }
 
-        // add the value to the mass balance equation
-        globalIndex const rowIndex = dofNumber[ei] - rankOffset;
-        localRhs[rowIndex] += rhsContributionArrayView[a] / sizeScalingFactor; // scale the contribution by the sizeScalingFactor here!
-      } );
+          // add the value to the mass balance equation
+          globalIndex const massRowIndex   = dofNumber[ei] - rankOffset;
+          globalIndex const energyRowIndex = massRowIndex + 1;
+          real64 const rhsValue = rhsContributionArrayView[a] / sizeScalingFactor; // scale the contribution by the sizeScalingFactor here!
+          localRhs[massRowIndex] += rhsValue; 
+          //add the value to the energey balance equation if the flux is positive (i.e., it's a producer)
+          if( rhsContributionArrayView[a] > 0.0 )
+          {
+            globalIndex const pressureDofIndex    = dofNumber[ei] - rankOffset;
+            globalIndex const temperatureDofIndex = pressureDofIndex + 1;
+            
+            localRhs[energyRowIndex] += enthalpy[ei][0] * rhsValue;
+
+            globalIndex dofIndices[2]{pressureDofIndex, temperatureDofIndex};
+            real64 jacobian[2]{rhsValue * dEnthalpy_dPressure[ei][0], rhsValue * dEnthalpy_dTemperature[ei][0]};
+
+            localMatrix.template addToRow< serialAtomic >( energyRowIndex,
+                                                           dofIndices,
+                                                           jacobian,
+                                                           2 );
+          }
```

Comment:
Here I recommend adding some logging (when logLevel is 1 or larger) to let the user know what the code does in the two configurations: 
- Thermal + producing source flux: a message saying that both the mass and energy equations in the cell are modified
- Thermal + injecting source flux: a message saying that only the mass equation in the cell is modified